pkg/cri/server/sandbox_stats_windows.go

```diff
@@ -19,21 +19,324 @@ package server
+	"github.com/Microsoft/hcsshim/hcn"
+	"github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/log"
+	containerstore "github.com/containerd/containerd/pkg/cri/store/container"
 	sandboxstore "github.com/containerd/containerd/pkg/cri/store/sandbox"
+	"github.com/containerd/containerd/pkg/cri/store/stats"
+	"github.com/containerd/typeurl"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
 )
 
-func (c *criService) podSandboxStats(ctx context.Context, sandbox sandboxstore.Sandbox, stats interface{}) (*runtime.PodSandboxStats, error) {
-	return nil, fmt.Errorf("pod sandbox stats not implemented on windows: %w", errdefs.ErrNotImplemented)
+func (c *criService) podSandboxStats(
+	ctx context.Context,
+	sandbox sandboxstore.Sandbox) (*runtime.PodSandboxStats, error) {
+	meta := sandbox.Metadata
+
+	if sandbox.Status.Get().State != sandboxstore.StateReady {
+		return nil, fmt.Errorf("failed to get pod sandbox stats since sandbox container %q is not in ready state", meta.ID)
+	}
+
+	podSandboxStats := &runtime.PodSandboxStats{}
+	podSandboxStats.Windows = &runtime.WindowsPodSandboxStats{}
+	podSandboxStats.Attributes = &runtime.PodSandboxAttributes{
+		Id:          meta.ID,
+		Metadata:    meta.Config.GetMetadata(),
+		Labels:      meta.Config.GetLabels(),
+		Annotations: meta.Config.GetAnnotations(),
+	}
+
+	metrics, containers, err := c.listWindowsMetricsForSandbox(ctx, sandbox)
+	if err != nil {
+		return nil, fmt.Errorf("failed to obtain container stats during podSandboxStats call: %w", err)
+	}
+	podCPU, containerStats, err := c.toWindowsStats(metrics, sandbox, containers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert container stats during podSandboxStats call: %w", err)
+	}
+	podSandboxStats.Windows.Cpu = podCPU.Cpu
+	podSandboxStats.Windows.Memory = podCPU.Memory
+	podSandboxStats.Windows.Containers = containerStats
+
+	podSandboxStats.Windows.Network = windowsNetworkUsage(ctx, sandbox)
+
+	pidCount, err := c.getSandboxPidCount(ctx, sandbox)
+	if err != nil {
+		return nil, err
+	}
+	timestamp := time.Now()
+	podSandboxStats.Windows.Process = &runtime.WindowsProcessUsage{
+		Timestamp:    timestamp.UnixNano(),
+		ProcessCount: &runtime.UInt64Value{Value: pidCount},
+	}
+
+	return podSandboxStats, nil
+}
+
+func (c *criService) toWindowsStats(metrics []*types.Metric, sandbox sandboxstore.Sandbox, containers []containerstore.Container) (*runtime.WindowsContainerStats, []*runtime.WindowsContainerStats, error) {
+	statsMap := make(map[string]*types.Metric)
+	for _, stat := range metrics {
+		statsMap[stat.ID] = stat
+	}
+
+	metric := statsMap[sandbox.ID]
+	s, err := typeurl.UnmarshalAny(metric.Data)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to extract container metrics: %w", err)
+	}
+
+	podStats := s.(*wstats.Statistics)
+	podRuntimeStats, err := c.windowsCPUAndMemoryStats(podStats, sandbox.Stats)
```

Comment:
Is `sandbox.Stats` nil or populated for HostProcess containers (which do not start a sandbox image)?

I just want to make sure that we'll report stats for all the containers in the pods that are running HostProcess containers since it looks like `appendCputPodStats()` and `appendMemoryPodStats()` both no-op if `podRuntimeStats` is nil.

Reply:
It returns a blank stats object https://github.com/microsoft/hcsshim/blob/main/cmd/containerd-shim-runhcs-v1/task_wcow_podsandbox.go#L300-L303

Reply:
Tested this out and it looks like the stats object is returned but the data is nil, so this is failing on `podStats := s.(*wstats.Statistics)`.  

Is there any way to determine that the podsandbox is host process here?  I am thinking the best approach would be, if `metric.Data` is nil then we can create a zero's stat object to begin appending the containers metrics too.

Reply:
I've updated this to handle the scenario where the stats are not present and it is now working with hostprocess pods